File: save_folder_files_to_markdown.js
// Generates a markdown file from files in the current directory,
// using rules to include or exclude specific file extensions and directories.
// It has some default rules that can be modified as needed.
// The generated markdown file will contain code blocks for each file.
// Usage: node save_folder_files_to_markdown.js

const fs = require('fs');
const path = require('path');

// Configuration
const OUTPUT_FILE_PATH = 'combined_files.md';
const FILE_EXTENSIONS_TO_INCLUDE = new Set([
  '.sh',
  '.py',
  '.js',
  '.ts',
  '.html',
  '.md',
  '.rst',
  '.json',
  '.yml',
  '.yaml',
]);
const ALLOWED_HIDDEN_DIRECTORIES = new Set(['.github', '.devcontainer']);
const DIRECTORIES_TO_SKIP = new Set(['node_modules']);
const FILES_TO_SKIP = new Set([OUTPUT_FILE_PATH, path.basename(__filename)]);
const SKIP_HIDDEN_DIRECTORIES = true;

const MARKDOWN_TEMPLATE = `
\`\`\`{relativePath}
{content}
\`\`\`

`;

const baseWorkingPath = process.cwd();

// Helper function to recursively walk through directories
function* walkSync(dir) {
  const files = fs.readdirSync(dir, { withFileTypes: true });
  for (const file of files) {
    const fullPath = path.join(dir, file.name);
    if (file.isDirectory()) {
      yield* walkSync(fullPath);
    } else {
      yield fullPath;
    }
  }
}

function shouldSkipDirectory(parts) {
  return parts.some((part) => DIRECTORIES_TO_SKIP.has(part));
}

function isHiddenDirectory(parts) {
  return parts.some(
    (part) => part.startsWith('.') && !ALLOWED_HIDDEN_DIRECTORIES.has(part)
  );
}

let bigString = '';

// Walk through the files
for (const filePath of walkSync(baseWorkingPath)) {
  const relativePath = path.relative(baseWorkingPath, filePath);
  const parts = relativePath.split(path.sep);

  if (
    shouldSkipDirectory(parts) ||
    (SKIP_HIDDEN_DIRECTORIES && isHiddenDirectory(parts))
  ) {
    continue;
  }

  if (FILES_TO_SKIP.has(path.basename(filePath))) {
    continue;
  }

  if (FILE_EXTENSIONS_TO_INCLUDE.has(path.extname(filePath))) {
    const content = fs.readFileSync(filePath, 'utf-8');
    bigString += MARKDOWN_TEMPLATE.replace(
      '{relativePath}',
      relativePath
    ).replace('{content}', content);
  }
}

// Write the output to the markdown file
fs.writeFileSync(OUTPUT_FILE_PATH, bigString, 'utf-8');

File: save_folder_files_to_markdown.py
"""
Generates a Markdown file from files in the current directory,
using rules to include or exclude specific file extensions and directories.
It has some default rules that can be modified as needed.
The generated markdown file will contain code blocks for each file.

Usage:
$ python save_folder_files_to_markdown.py
"""

OUTPUT_FILE_PATH = "combined_files.md"
FILE_EXTENSIONS_TO_INCLUDE = {'.sh', '.py', '.js', '.ts', '.html', '.md', '.rst', '.json', '.yml', '.yaml'}
ALLOWED_HIDDEN_DIRECTORIES = {'.github', '.devcontainer'}
DIRECTORIES_TO_SKIP = {'node_modules'}
FILES_TO_SKIP = {OUTPUT_FILE_PATH, __file__}
SKIP_HIDDEN_DIRECTORIES = True

from pathlib import Path

MARKDOWN_TEMPLATE = """
```{relative_path}
{content}
```

"""

base_working_path = Path.cwd()

def should_skip_directory(path_parts):
    """Check if the path contains a directory that should be skipped."""
    return any(part in DIRECTORIES_TO_SKIP for part in path_parts)

def is_hidden_directory(path_parts):
    """Check if the path contains a hidden directory that should be skipped."""
    return any(
        part.startswith('.') and part not in ALLOWED_HIDDEN_DIRECTORIES 
        for part in path_parts
    )

big_string = ""

for path in base_working_path.rglob('*'):
    if should_skip_directory(path.parts):
        continue
    
    if SKIP_HIDDEN_DIRECTORIES and is_hidden_directory(path.parts):
        continue

    if path.name in FILES_TO_SKIP:
        continue

    if path.suffix in FILE_EXTENSIONS_TO_INCLUDE:
        relative_path = path.relative_to(base_working_path)
        content = path.read_text(encoding='utf-8')
        # Adds markdown code block tags with relative path:
        big_string += MARKDOWN_TEMPLATE.format(**locals())

with open(OUTPUT_FILE_PATH, "w", encoding="utf-8") as file:
    file.write(big_string)
